### DISCUSSION TITLE
sharkdp/fd is an optional dependency. What is it used for?

### DISCUSSION THREAD

**Discussion by astier** [nvim-telescope/telescope.nvim#1349] [Q&A] [answered]
At the beginning of the README it is mentioned that https://github.com/sharkdp/fd is an optional dependency. If I install it will it be used as the backend to find files for the find_files command instead of a telescope-native solution? Is it faster? What are the benefits? I installed fd and couldn't notice any difference except that `checkhealth` now says fd could be found.

**Comment by rodamaral** [comment, marked as answer]
> If I install it will it be used as the backend to find files for the find_files command instead of a telescope-native solution?

Telescope exposes the `find_command` option that allows you to pass an arbitrary shell command (with arguments) that you'd use in your shell to search for files. If this option is set, Telescope uses it. If not, Telescope will search your path for some executable names, as follows: _fd_, _fdfind_, _rg_, _find_ and _where_. If some of those executables exist, Telescope will create a full command using it as the base.

Every Linux distribution has a command called [find](https://man7.org/linux/man-pages/man1/find.1.html) for that and `fd` is a modern alternative.

Implementation at:
https://github.com/nvim-telescope/telescope.nvim/blob/978366ba46e2306b60ef64bd9e51a6c758fe48f0/lua/telescope/builtin/files.lua#L156
 
 ---
 
> Is it faster? What are the benefits?

Yes, against find. There're benchmarks [here](https://github.com/sharkdp/fd#benchmark). Another advantage is that fd can respect your gitignored folders and skip them. You can also create a `.fdignore` file for extra ignores.

---

> I installed fd and couldn't notice any difference except that checkhealth now says fd could be found.

Performance will start to be noticeable only in large repositories or folders.
fd is the default choice when found, but to be certain it is being used, set

```lua
pickers.find_files.find_command = {
        'fd',
        '--type',
        'f',
        -- '--no-ignore-vcs', respect your gitignored files?
        '--color=never',
        '--hidden',
        '--follow',
}
```

**Comment by astier** [comment]
Thank you for the detailed explanation.
